nixos/modules/virtualisation/digital-ocean-config.nix

```diff
@@ -0,0 +1,167 @@
+        initrd.kernelModules = [ "virtio_scsi" ];
+        kernelModules = [ "virtio_pci" "virtio_net" ];
+        loader = {
+          grub.device = "/dev/sda";
+          timeout = 0;
+          grub.configurationLimit = 0;
+        };
+      };
+      services.openssh = {
+        enable = true;
+        permitRootLogin = "prohibit-password";
+        passwordAuthentication = mkDefault false;
+      };
+      networking = {
+        firewall.allowedTCPPorts = [ 22 ];
+        hostName = mkDefault ""; # use Digital Ocean metadata server
+        useNetworkd = true;
+      };
+
+      /* Fetch the root password from the digital ocean metadata.
+       * There is no specific route for this, so we use jq to get
+       * it from the One Big JSON metadata blob */
+      systemd.services.digitalocean-set-root-password = {
+        enable = cfg.setRootPassword;
+        path = [ pkgs.curl pkgs.shadow pkgs.jq ];
+        description = "Set root password provided by Digitalocean";
+        wantedBy = [ "multi-user.target" ];
+        script = ''
+          set -e
+          ROOT_PASSWORD=$(curl --retry-connrefused http://169.254.169.254/metadata/v1.json | jq -r '.auth_key')
+          echo "root:$ROOT_PASSWORD" | chpasswd
+          '';
+        unitConfig = {
+          After = [ "network-online.target" ];
+          Wants = [ "network-online.target" ];
+        };
+        serviceConfig = {
+          Type = "oneshot";
+        };
+      };
+
+      /* Set the hostname from Digital Ocean, unless the user configured it in
+       * the NixOS configuration */
+      systemd.services.digitalocean-set-hostname = {
+        enable = hostName == "";
+        path = [ pkgs.curl pkgs.nettools ];
+        description = "Set hostname provided by Digitalocean";
+        wantedBy = [ "multi-user.target" ];
+        script = ''
+          set -e
+          DIGITALOCEAN_HOSTNAME=$(curl --retry-connrefused http://169.254.169.254/metadata/v1/hostname)
+          hostname $DIGITALOCEAN_HOSTNAME
+        '';
+        unitConfig = {
+          After =  [ "network-online.target" ];
+          Wants = [ "network-online.target" ];
+        };
+        serviceConfig = {
+          Type = "oneshot";
+        };
+      };
+
+      /* Fetch the ssh keys for root from Digital Ocean */
+      systemd.services.digitalocean-ssh-keys = {
+        enable = cfg.setSshKeys;
+        description = "Set root ssh keys provided by Digital Ocean";
+        wantedBy = [ "multi-user.target" ];
+        path = [ pkgs.curl ];
+        script = ''
+          set -e
+          mkdir -m 0700 -p /root/.ssh
+          curl --retry-connrefused --output /root/.ssh/authorized_keys http://169.254.169.254/metadata/v1/public-keys
+          chmod 600 /root/.ssh/authorized_keys
+        '';
+        serviceConfig = {
+          Type = "oneshot";
+          RemainAfterExit = true;
+        };
+        unitConfig = {
+          ConditionFileExists = "!/root/.ssh/authorized_keys";
```

Comment:
Doesn't this mean, we fetch public keys only once, and can never update them by updating the cloud-init metadata?

Reply:
Yes, but this is in line with behaviour of other images. hence  I copied it https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/ec2-data.nix#L29

Reply:
IIRC cloud-init is only really meant to be run once, I think you can override that behaviour if you really want to but many providers don't even allow you to modify the metadata. Does DO even have a way to update them?

Reply:
fair enough, the amazon cloud-init foo also only does it once.